internal/controllers/addon/monitoring_stack_reconciler.go

```diff
@@ -0,0 +1,222 @@
+	"fmt"
+	"strings"
+
+	monv1 "github.com/rhobs/obo-prometheus-operator/pkg/apis/monitoring/v1"
+	obov1alpha1 "github.com/rhobs/observability-operator/pkg/apis/monitoring/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8sApiErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+	"github.com/openshift/addon-operator/internal/controllers"
+)
+
+const MONITORING_STACK_RECONCILER_NAME = "monitoringStackReconciler"
+
+var errMonitoringStackSpecNotFound = fmt.Errorf("monitoring stack spec not found")
+
+type monitoringStackReconciler struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+func (r monitoringStackReconciler) Name() string {
+	return MONITORING_STACK_RECONCILER_NAME
+}
+
+func (r *monitoringStackReconciler) Reconcile(ctx context.Context,
+	addon *addonsv1alpha1.Addon) (ctrl.Result, error) {
+
+	// ensure creation of MonitoringStack object
+	latestMonitoringStack, err := r.ensureMonitoringStack(ctx, addon)
+	if err != nil {
+		if errors.Is(err, errMonitoringStackSpecNotFound) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// propagate the recently reconciled (created/updated) monitoring stack's status to the owner Addon
+	if monitoringStackAvailable := r.propagateMonitoringStackStatusToAddon(latestMonitoringStack, addon); !monitoringStackAvailable {
+		return handleExit(resultRetry), nil
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *monitoringStackReconciler) ensureMonitoringStack(ctx context.Context,
+	addon *addonsv1alpha1.Addon) (*obov1alpha1.MonitoringStack, error) {
+	if !HasMonitoringStack(addon) {
+		return nil, errMonitoringStackSpecNotFound
+	}
+	// create desired MonitoringStack
+	desiredMonitoringStack, err := r.getDesiredMonitoringStack(ctx, addon)
+	if err != nil {
+		return nil, err
+	}
+
+	// returns observed MonitoringStack object
+	reconciledMonitoringStack, err := r.reconcileMonitoringStack(ctx, desiredMonitoringStack)
+	if err != nil {
+		return nil, err
+	}
+	return reconciledMonitoringStack, nil
+}
+
+func (r *monitoringStackReconciler) propagateMonitoringStackStatusToAddon(monitoringStack *obov1alpha1.MonitoringStack, addon *addonsv1alpha1.Addon) (monitoringStackStackAvailable bool) {
+	availableCondition, reconciledCondition := obov1alpha1.Condition{}, obov1alpha1.Condition{}
+	availableConditionFound, reconciledConditionFound := false, false
+
+	// iterate until all conditions have been traversed or both `available` and `reconcile` conditions have been found
+	for i := 0; (i < len(monitoringStack.Status.Conditions)) && !(availableConditionFound && reconciledConditionFound); i++ {
```

Comment:
If they'd used `k8s.io/apimachinery/pkg/apis/meta/v1` we could have used `meta.FindStatusCondition()`  🥲

Reply:
Yeah you're right :(